dotnet/src/Planners/Planners.OpenAI/Stepwise/FunctionCallingStepwisePlanner.cs

```diff
@@ -0,0 +1,309 @@
+        string systemMessage = await this._promptTemplateFactory.Create(this._initialPlanPrompt, new PromptTemplateConfig()).RenderAsync(systemContext, cancellationToken).ConfigureAwait(false);
+
+        chatHistory.AddSystemMessage(systemMessage);
+        chatHistory.AddUserMessage(goal);
+
+        return chatHistory;
+    }
+
+    private async Task<ChatHistory> BuildChatHistoryForStepAsync(
+        string goal,
+        string initialPlan,
+        CancellationToken cancellationToken)
+    {
+        var chatHistory = this._chatCompletion.CreateNewChat();
+
+        // Add system message with context about outputs of previous functions
+        var systemContext = this._kernel.CreateNewContext();
+        systemContext.Variables.Set(GoalKey, goal);
+        systemContext.Variables.Set(InitialPlanKey, initialPlan);
+        var systemMessage = await this._promptTemplateFactory.Create(this._stepPrompt, new PromptTemplateConfig()).RenderAsync(systemContext, cancellationToken).ConfigureAwait(false);
+
+        chatHistory.AddSystemMessage(systemMessage);
+
+        return chatHistory;
+    }
+
+    #region private
+
+    /// <summary>
+    /// The configuration for the StepwisePlanner
+    /// </summary>
+    private FunctionCallingStepwisePlannerConfig Config { get; }
+
+    // Context used to access the list of functions in the kernel
+    private readonly IKernel _kernel;
+    private readonly IChatCompletion _chatCompletion;
+    private readonly ILogger? _logger;
+
+    /// <summary>
+    /// The prompt (system message) used to generate the initial set of steps to perform.
+    /// </summary>
+    private readonly string _initialPlanPrompt;
+
+    /// <summary>
+    /// The prompt (system message) for performing the steps.
+    /// </summary>
+    private readonly string _stepPrompt;
+
+    /// <summary>
+    /// The prompt renderer to use for the system step
+    /// </summary>
+    private readonly BasicPromptTemplateFactory _promptTemplateFactory;
+
+    /// <summary>
+    /// The name to use when creating semantic functions that are restricted from plan creation
+    /// </summary>
+    private const string RestrictedPluginName = "OpenAIFunctionsStepwisePlanner_Excluded";
+
+    /// <summary>
+    /// The user message to add to the chat history for each step of the plan.
+    /// </summary>
+    private const string StepwiseUserMessage = "Perform the next step of the plan if there is more work to do. When you have reached a final answer, use the UserInteraction_SendFinalAnswer function to communicate this back to the user.";
+
+    // Context variable keys
+    private const string AvailableFunctionsKey = "available_functions";
+    private const string InitialPlanKey = "initial_plan";
+    private const string GoalKey = "goal";
+
+    #endregion private
+
+    /// <summary>
+    /// Contains helper functions used by the <see cref="FunctionCallingStepwisePlanner"/>
+    /// </summary>
+    public class UserInteraction
```

Comment:
Should this be internal and/or sealed?

Reply:
Yes to sealed, but it has to be public to be registered with the kernel / included in the function list.